cluster-api/machines-api/proposal.md

```diff
@@ -0,0 +1,164 @@
+Minimalistic Machines API
+=========================
+
+This proposal is for a minimalistic start to a new Machines API, as part of the
+overall Cluster API project. It is intended to live outside of core Kubernetes
+and add optional machine management features to Kubernetes clusters.
+
+## Capabilities
+
+This API strives to be able to add these capabilities:
+
+1. A new Node can be created in a declarative way, including Kubernetes version
+   and container runtime version. It should also be able to specify
+   provider-specific information such as OS image, instance type, disk
+   configuration, etc., though this will not be portable.
+
+2. A specific Node can be deleted, freeing external resources associated with
+   it.
+
+3. A specific Node can have its kubelet version upgraded or downgraded in a
+   declarative way\*.
+
+4. A specific Node can have its container runtime changed, or its version
+   upgraded or downgraded, in a declarative way\*.
+
+5. A specific Node can have its OS image upgraded or downgraded in a declarative
+   way\*.
+
+\*  It is an implementation detail of the provider if these operations are
+performed in-place or via Node replacement.
+
+## Proposal
+
+This proposal introduces a new API type: Machine. See the full definition in
+[types.go](types.go).
+
+A "Machine" is the declarative spec for a Node, as represented in Kuberenetes
+core. If a new Machine object is created, a provider-specific controller will
+handle provisioning and installing a new host to register as a new Node matching
+the Machine spec. If the Machine's spec is updated, a provider-specific
+controller is responsible for updating the Node in-place or replacing the host
+with a new one matching the updated spec. If a Machine object is deleted, the
+corresponding Node should have its external resources released by the
+provider-specific controller, and should be deleted as well.
+
+Fields like the kubelet version, the container runtime to use, and its version,
+are modeled as fields on the Machine's spec. Any other information that is
+provider-specific, though, is part of an opaque ProviderConfig string that is
+not portable between different providers.
+
+The ProviderConfig is recommended to be a serialized API object in a format
+owned by that provider, akin to the [Component Config](https://goo.gl/opSc2o)
+pattern. This will allow the configuration to be strongly typed, versioned, and
```

Comment:
I personally think this is a little bit of a cheat, and I'd like us to avoid this if possible, but I also recognize that we'll always need an escape hatch in the long term.  My concern in the short/medium term is that allowing this means we avoid defining common fields where we actually could do so.

Reply:
I absolutely, 100% agree that it's a simplification for the short-term. I'll be fully transparent here:

My other proposal was a little too over-indexed on trying to support a cloud-agnostic cluster autoscaler as an initial customer. That was appealing from a design standpoint, because the cluster autoscaler already exists and actually has concrete requirements to build off of, but also seemed like a logical thing you would want to do with a declarative NodeSet concept. After a lot of discussions, though, the feedback was:

1. For a very first proposal of a brand-new API, it had a lot of new concepts to grasp.

To be fair, I had new types across two dimensions: Machine / MachineSet / MachineDeployment in one dimension, and Machine / MachineClass / MachineTemplate across the other. I was urged to think about the absolute minimum we could get away with for the first iteration, and only add more complexity and new concepts later when we were certain they were necessary.

My other design also had the requirement that in order to create even a single simple node, you had to first create a cloud-specific MachineTemplate, then register it as a MachineClass, and then finally create a Machine that referenced the class. It seemed like a lot of overhead just to support a single custom node that you had no intent to reuse. So, one design principle that emerged from that was that if I just want to create a single Machine and not care about portability, I should be able to. I think there's still room to evolve the API and introduce other concepts down the road, but it seems reasonable to me that if you don't care about portability, it should be possible to create a single custom Machine without the overhead of needing to use other concepts, which means that we'll need something akin to the opaque ProviderConfig blob we have now to actually be able to feed the right values into the cloud-specific node creation APIs.

2. Rebasing the cluster autoscaler on top of our new APIs isn't actually delivering any new value, since cluster autoscaling already exists today.

It's a nice-to-have down the road, but in order to get the momentum we want for the project, we should focus first on the new functionality this enables that was never possible before (like cloud- and deployment-agnostic cluster upgrades). If you remove cluster autoscaling (and node autoprovisioning) as a client, at least for the short term, then I think most of the benefits of having more cloud-agnostic fields in the MachineSpec disappear as well.

I could definitely be wrong here, and welcome more feedback. But, the intent of this proposal was "what is the absolute bare minimum that we can all agree is a good starting point?" and then continue to add more as we see appropriate.

Some of the things I had considered for inclusion in the MachineSpec:

* OS image
  * This could be represented as a single string across most providers, but the values wouldn't be portable anyway. If the values are cloud-specific, then moving the entire field to be cloud-specific means that each provider can represent this however it would like. For instance, in GCE, OS images are actually more naturally represented in a structured way, since they have a project, family, name, etc. In DigitalOcean, OS Images can also be referred to via int IDs, so an IntOrStr might be more appropriate there.
* Disk configuration
  * I don't know if I personally have enough of a grasp on how much power is needed to be useful here. I've swung back and forth between two extremes: a simplified view of just specifying a total amount of working space desired, and on the other end of the spectrum, having a full array of structs to represent disks, along with fields for how much capacity each disk should have, whether it should be bootable, etc. At this point, deferring disk setup to the ProviderConfig lets us look at how different early adopters represent this in their config, and we can always bubble it back up to a generic representation in MachineSpec in the future once we think we have a grasp of where we want to end up on the power/usability trade-off scale.
* Preemptible
  * AWS and GCE support these, but I think Azure only supports low-priority VMs in Batch, and I don't know if this concept exists in other on-premise environments like vSphere.
* Topology
  * Like OS Image, these are similar-enough concepts across environments (regions, zones, availability zones, etc), but I couldn't see the actual values being portable at all. Also, the number of dimensions you might want to support in different environments might differ wildly. In some clouds, a single availability zone might be enough to specify. If you're on-premise, you might want a lot of custom fields like datacenter, rack, etc.

There are many more, but you get the point. Are there particular concepts that you feel we should represent in the ClusterSpec now (or in the near term)?

Reply:
@justinsb it shouldn't be a problem to lift common fields into this API as we notice them. Providers can always extend their controllers to check for the API field and fall back on their old provider-specific field.

Reply:
@pipejakob After re-watching all meeting recordings, I see the justification for using serialized blob - simplicity. I'm 100% for simplicity. However I think that `providerConfig` is very similar to `env` in `ContainerSpec`. There you can define a literal value, when you don't need reusability  and reference, when you need it. Something like

```yaml
providerConfig:
  value: >
    {
      "apiVersion": "gceproviderconfig/v1alpha1",
      "kind": "GCEProviderConfig",
      "machineType": "n1-standard-1",
      ...
    }
  valueRef:
    apiVersion: gceproviderconfig/v1alpha1
    kind: GCEProviderConfig
    name: config-1
```
can be suitable for both use-cases and it feels very similar to what people know and use in `Pods`

p.s. this is a re-post of a previous comment

Reply:
@pipejakob After re-watching all meeting recordings, I see the justification for using serialized blob - simplicity. I'm 100% for simplicity. However I think that `providerConfig` is very similar to `EnvVar` in `Container`. There you can define a literal value, when you don't need reusability (or doing it quickly) and reference, when you need it. Something like

```yaml
providerConfig: # either value or valueRef
  value: >
    {
      "apiVersion": "gceproviderconfig/v1alpha1",
      "kind": "GCEProviderConfig",
      "machineType": "n1-standard-1",
      ...
    }
  valueRef:
    apiVersion: gceproviderconfig/v1alpha1
    kind: GCEProviderConfig
    name: config-1
```

can be suitable for both use-cases and it feels very similar to what people know and use in `Pods`

Edit:
even better it would be if have:

```yaml
providerConfig:
  value: # either value (runtime.Object/runtime.RawExtension) or valueRef
    apiVersion: gceproviderconfig/v1alpha1,
    kind: GCEProviderConfig,
    machineType: n1-standard-1,
    project: test-1-test
    ...
  valueRef:
    apiVersion: gceproviderconfig/v1alpha1
    kind: GCEProviderConfig
    name: config-1
```

p.s. this is a re-post of a previous comment (wrong user)

Reply:
Why not using `runtime.Object/runtime.RawExtension` instead of `string`?

Reply:
I second this. runtime.RawExtension needs an apiVersion and will probably fall back to `Unstructured` (@mrIncompetent can you confirm that?). Alternatively, a generic JSON type will do, compare https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go#L28.

Reply:
@sttts `runtime.RawExtension.Object` would fall back to `runtime.Unknown`.
Though i've been using it without `apiVersion` and more like https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go#L28
So i only used `runtime.RawExtension.Raw`. In this case, `RawExtension.Object` will be `nil`

Reply:
@mvladev @pipejakob @justinsb @sttts
Any objections on using `runtime.RawExtension` for `providerConfig`

Reply:
sounds good

Reply:
No objection, it's way better than `string`.